xray/services/scan.go

```diff
@@ -189,6 +190,34 @@ type GraphNode struct {
 	Parent *GraphNode `json:"-"`
 }
 
+// FlattenGraph creates a set of dependencies from the given graph, and a map of the paths to each dependency.
+func FlattenGraph(graph []*GraphNode) []*GraphNode {
+	flatGraph := []*GraphNode{}
+	depsSet := datastructures.MakeSet[string]()
+	for _, node := range graph {
+		flatGraph = append(flatGraph, getUniqueDependencies(node, depsSet)...)
+	}
+	return []*GraphNode{{Id: "root", Nodes: flatGraph}}
+}
+
+func getUniqueDependencies(node *GraphNode, depsSet *datastructures.Set[string]) []*GraphNode {
```

Comment:
Wouldn't work better with a map of string -> *GraphNode?
The key is the dependency ID and the value is the GraphNode. No return value.

Possible implementation (please double check me):
```go
func FlattenGraph(graph []*GraphNode) []*GraphNode {
	allDependencies := map[string]*GraphNode{}
	for _, node := range graph {
		populateUniqueDependencies(node, allDependencies)
	}
	return []*GraphNode{{Id: "root", Nodes: maps.Values(allDependencies)}}
}

func populateUniqueDependencies(node *GraphNode, allDependencies map[string]*GraphNode) {
	if _, exist := allDependencies[node.Id]; exist {
		return
	}
	allDependencies[node.Id] = &GraphNode{Id: node.Id}
	for _, dependency := range node.Nodes {
		populateUniqueDependencies(dependency, allDependencies)
	}
}
```

Reply:
Nice! Didn't know about the `maps.Values()`. Works great.